### PR TITLE
Fix the ConcurrentModificationException on the _trackEventsOutsideAltSlaMap in the EventProducer

### PR DESCRIPTION
**Pull request by somandal**
On enabling the additional warning log to track TopicPartitions seeing events outside alternate SLA, we saw an instance of STALLED partitions. Further investigation showed that the logging function in the EventProducer was throwing a ConcurrentModificationException under certain circumstances. When this occurs, the caller's onCompletion send callback isn't called. This can cause an issue where we never acknowledge the successful send, and the BMM flushless producer never commits a newer offset.

This PR fixes this in two ways:
- Make the _trackEventsOutsideAltSlaMap a ConcurrentHashMap
- Catch all exceptions that can be thrown by hashmap operations and log them rather than throwing.

Important: DO NOT REPORT SECURITY ISSUES DIRECTLY ON GITHUB.  
For reporting security issues and contributing security fixes,  
please, email security@linkedin.com instead, as described in  
the contribution guidelines.

Please, take a minute to review the contribution guidelines at:  
https://github.com/linkedin/Brooklin/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java [hunk 1]
```diff
@@ -272,20 +273,28 @@ private void performSlaRelatedLogging(DatastreamRecordMetadata metadata, long ev
...
+        long timeSinceLastLog = System.currentTimeMillis() - _lastEventsOutsideAltSlaLogTimeMs;
+        if (timeSinceLastLog >= _numEventsOutsideAltSlaFrequencyMs) {
+          _trackEventsOutsideAltSlaMap.forEach((topicPartition, numEvents) -> _logger.warn(
+              "{} had {} event(s) with latency greater than alternate SLA of {} ms in the last {} ms for "
+                  + "datastream {}", topicPartition, numEvents, _availabilityThresholdAlternateSlaMs, timeSinceLastLog,
+              getDatastreamName()));
+          _trackEventsOutsideAltSlaMap.clear();
+          _lastEventsOutsideAltSlaLogTimeMs = System.currentTimeMillis();
+        }
+      } catch (ConcurrentModificationException | NullPointerException | IllegalArgumentException | ClassCastException
```
Comment (vmaheshw) [inline datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java:292]: ConcurrentHashMap will not throw ConcurrentModificationException. You can get rid of it.

**Review by vmaheshw — COMMENTED** [review]
Do you think there is a value in moving sendCallback.onCompletion to finally block in `onSendCallback`, to avoid any future bug in `checkpoint()` or `reportMetrics` does not impact the checkpoint. The reason is this is a critical piece for FlushlessEventProducer otherwise we will be sending duplicate data, till the issue is identified and fixed.

File: datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java (hunk 1 above)
Comment (somandal) [reply datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java]: done

**Comment by somandal** [comment]
> Do you think there is a value in moving sendCallback.onCompletion to finally block in `onSendCallback`, to avoid any future bug in `checkpoint()` or `reportMetrics` does not impact the checkpoint. The reason is this is a critical piece for FlushlessEventProducer otherwise we will be sending duplicate data, till the issue is identified and fixed.

As discussed offline, I've modified the code to catch all exceptions and wrap them in the sendFailureException and call the onCompletion in the finally block.

**Review by vmaheshw — APPROVED** [review]

**Review by jzakaryan — APPROVED** [review]

**Pull request merged**